hypha/apply/funds/workflow.py

```diff
@@ -265,6 +269,7 @@ def make_permissions(edit=None, review=None, view=None):
                 "almost": _("Accept but additional info required"),
                 "accepted": _("Accept"),
                 "rejected": _("Dismiss"),
+                "withdrawn": _("Withdraw"),
```

Comment:
I thought (but did not test) that the transitions list was what allowed step A to go to step B, namely it's allowed if step B shows up in step A's transition list. But maybe I am missing the point here, looking more closely.

Reply:
Transitions are the options that staff gets when they want to change status.

Reply:
So when I remove `...withdraw...` from the transitions, there is no effect when I apply and then click withdraw. But this is probably due to the implementation of the `withdraw` method in `SubmissionWithdrawView` where it explicitly looks through the `transitions`: `            for action in obj.workflow[obj.status].transitions.keys()`. Should that `for` loop go away and the `if len(withdraw_actions)` condition go away as well, and do a different kind of lookup to find the action?

Reply:
Hmm. When I remove all the `...withdraw...` from the transitions and change the `SubmissionWithdrawView.withdraw` implementation to this (starting at existing `withdraw_actions`):
```python

        if len(withdraw_actions) == 1:
            action = withdraw_actions[0]
            obj.perform_transition(
                action, self.request.user, request=self.request, notify=False
            )
        elif len(withdraw_actions) > 1:
            raise PermissionDenied(f'In workflow "{obj.workflow}" too many withdraw actions: "{withdraw_actions}"')
        elif len(withdraw_actions) < 1:
            raise PermissionDenied(f'No withdraw actions found in workflow "{obj.workflow}"')
```
then the call to `obj.perform_transition` fails on line 347.

```
  File "/home/everybody/code/ots/ext/hypha-clean/hypha/apply/funds/views.py", line 1611, in withdraw
    obj.perform_transition(
  File "/home/everybody/code/ots/ext/hypha-clean/hypha/apply/funds/models/submissions.py", line 347, in perform_transition
    raise PermissionDenied(f'Invalid "{ action }" transition')
django.core.exceptions.PermissionDenied: Invalid "withdrawn" transition
```

Unless there's another way to advance a workflow, I don't see how this can work. What am I missing?

Reply:
My mistake, the transitions are also used to check if the status change is allowed.

So they need to be there but I think we can add custom permissions to them so only applicant are allowed to use them. Example below.

https://github.com/HyphaApp/hypha/blob/f63f0b2a2ea32ed7e4b982f479ebdf7dcc15c0ff/hypha/apply/funds/workflow.py#L402-L408

Reply:
The permissions are already checked twice elsewhere, so I don't think we need a third.

Check one is in the `SubmissionWithdrawView.withdraw` function and check two is whenever any permissions object in `workflow.py`  is checked because `make_permissions` in `workflow.py` defaults `"withdraw"` to `[applicant_can]`. Yes?

Reply:
This permission hide this status from staff so they do not see it and by accident make use of it.

Reply:
I also think we should limit "withdraw" to `applicant_edit_permissions`. 

```
applicant_edit_permissions = make_permissions(
    edit=[applicant_can, partner_can], review=[staff_can], withdraw=[applicant_can]
)
``` 

```
def make_permissions(edit=None, review=None, view=None, withdraw=None):
    return {
        "edit": edit or [],
        "review": review or [],
        "view": view
        or [
            staff_can,
            applicant_can,
            reviewer_can,
            partner_can,
        ],
        "withdraw": withdraw or [],
    }
```

Reply:
> This permission hide this status from staff so they do not see it and by accident make use of it.

Without the suggested change (i.e. while on 5b656bbc2e89c9d93f7ab90e690783667c852d87) locally, I logged in as staff and could not see the Withdraw button. Very well, that's because of a check on the page not displaying a link. So I logged in as the applicant, then copied the "withdraw" link, logged out, logged in as staff, and pasted the withdraw link, clicked the button: `PermissionDenied`:

```
hypha-clean/hypha/apply/funds/views.py", line 1599, in withdraw
    raise PermissionDenied
django.core.exceptions.PermissionDenied
```

What code is this? The new view:
```python
class SubmissionWithdrawView(SingleObjectTemplateResponseMixin, BaseDetailView):
#...
   def withdraw(self, request, *args, **kwargs):
#...
        if not request.user.is_applicant:
            raise PermissionDenied #here is line 1599.
```

Would you prefer these two lines in the view to be removed so that the permissions are properly concentrated in the one area?

Reply:
@bickelj As staff check the "Update status" button/dialog. There staff will see "Withdraw" unless permission is set to hide it.

Reply:
@frjo I made a lot of noise above which you can read as "the permissions model for workflows is not clear or not working as I would expect on first glance." This is particularly around [the suggestion above](https://github.com/HyphaApp/hypha/pull/3298#discussion_r1696513386) which seemed straightforward to me. I used this snippet for all the `"withdrawn"` keys in `transitions` but somehow staff can still make the transition to withdrawn:
```python

                "withdrawn": {
                    "display": _("Withdraw"),
                    "permissions": {UserPermissions.APPLICANT},
                    "method": "withdraw",
                }
```

Reply:
I think this is resolved for all practical purposes.